New upload algorithm and modified actions

lukaskremla opened the pull request:
Added new upload algorithm logic for building target file paths and cleaned up existing methods.

Execute in REPL action is now only clickable for single non-directory file selections.

Upload action now uses the new upload algorithm and its description changes based on the current file/folder selection.

lukaskremla committed:
Fixed and improved upload algorithm and actions

lukaskremla commented:
Regarding the execute in REPL, in the old PRs we have at some point discussed whether the plugin should allow users to run non `.py` files in REPL. 

You said that it should be possible, but I think that this functionality should be adjusted to only allow `.py` and `.mpy` files to be executed in REPL. Just like Windows won't let me execute an executable file if I change its extension from `.exe` to `.txt `this plugin shouldn't let people execute other files. I don't see a use case where I'd put a MicroPython script inside of a file with a `.txt` extension and then need to execute it in REPL.

As a user of this plugin it seems very reasonable that this plugin will only allow me to execute `.py` or `.mpy` files in REPL.

elmot commented on src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:
```diff
@@ -259,28 +262,55 @@ class OpenMpyFile : ReplAction("Open file", true) {
+            }
+
+            if (normalFileCount >= 1 && directoryCount >= 1) {
+                e.presentation.text = "Upload Items to MicroPython Device"
```
This code is hard to read, and there are six branches for five cases.
probably it's easier to write like. Also normalFileCound sounds a bit misleading
```kotlin
if(fileCount  == 0) {
  if(directoryCount == 1) {...} else {...}
} else if(directoryCount == 0) {
  if(fileCount == 1) {...} else {...}
} else {...}
```

lukaskremla replied:
I agree with you on renaming it to `fileCount` but are you sure that the existing conditional is too hard to read? It seems more readable to me than nesting them.

The last else condition might be unnecessary, I just left it there as a placeholder just in case.

elmot replied:
four ifs in IMO two layers of ifs with no boolean operators, four conditions in total is easier than  five ifs with six conditions

lukaskremla replied:
I've looked at it again, agreed.

lukaskremla committed:
renamed normalFileCount to fileCount

elmot commented:
Now the PR is not compatible with the base branch and needs to be rebased

lukaskremla committed:
Rebased the new upload algorithm approach and actions

lukaskremla committed:
Merge remote-tracking branch 'origin/New-upload-algorithm-and-modified-actions' into New-upload-algorithm-and-modified-actions

# Conflicts:
#	src/main/kotlin/com/jetbrains/micropython/nova/actions.kt
#	src/main/kotlin/com/jetbrains/micropython/run/MicroPythonRunConfiguration.kt

lukaskremla committed:
Removed "." from the extension comparison

lukaskremla commented:
> Now the PR is not compatible with the base branch and needs to be rebased

Should be up to date again. I've modified the new upload algorithm to hopefully include all of the important changes you've made.

I have also added the extension check back as mentioned. I did the python file comparison the way you suggested before, but I believe that the mpy comparison can't be done similarly. 

I think the check should be present as I have explained above, but ultimately this is up to you.
